pkg/networkservice/common/excludedprefixes/server.go

```diff
@@ -87,15 +88,56 @@ func (eps *excludedPrefixesServer) Request(ctx context.Context, request *network
 	if conn.GetContext().GetIpContext() == nil {
 		conn.Context.IpContext = &networkservice.IPContext{}
 	}
-	prefixes := eps.prefixPool.Load().(*ippool.PrefixPool).GetPrefixes()
-	log.FromContext(ctx).Debugf("ExcludedPrefixesService: adding excluded prefixes to connection: %v", prefixes)
-	ipCtx := conn.GetContext().GetIpContext()
-	ipCtx.ExcludedPrefixes = removeDuplicates(append(ipCtx.GetExcludedPrefixes(), prefixes...))
+	// excluded prefixes from file
+	prefixesFromFile := eps.prefixPool.Load().(*ippool.PrefixPool).GetPrefixes()
+	// excluded prefixes from client
+	clientPrefixes := conn.GetContext().GetIpContext().GetExcludedPrefixes()
+
+	sort.Strings(prefixesFromFile)
+	sort.Strings(clientPrefixes)
+
+	if IsEqual(prefixesFromFile, clientPrefixes) {
+		return next.Server(ctx).Request(ctx, request)
+	}
+
+	prefixesInfo, _ := load(ctx)
+
+	clientPrefixesChanged := !IsEqual(prefixesInfo.previousClientPrefixes, clientPrefixes)
+	filePrefixesChanged := !IsEqual(prefixesInfo.previousFilePrefixes, prefixesFromFile)
+
+	finalPrefixes := clientPrefixes
+	if !clientPrefixesChanged && filePrefixesChanged {
+		finalPrefixes = removePrefixes(clientPrefixes, prefixesInfo.previousDiff)
+	}
+
+	finalPrefixes = removeDuplicates(append(finalPrefixes, prefixesFromFile...))
+	request.GetConnection().GetContext().GetIpContext().ExcludedPrefixes = finalPrefixes
+
+	if filePrefixesChanged {
+		prefixesInfo.previousFilePrefixes = make([]string, len(prefixesFromFile))
+		copy(prefixesInfo.previousFilePrefixes, prefixesFromFile)
+	}
+
+	if clientPrefixesChanged {
+		prefixesInfo.previousClientPrefixes = make([]string, len(finalPrefixes))
+		copy(prefixesInfo.previousClientPrefixes, finalPrefixes)
+	}
+
+	prefixesInfo.previousDiff = Subtract(clientPrefixes, prefixesFromFile)
+
+	log.FromContext(ctx).
+		WithField("excludedPrefixes", "server").
+		WithField("prefixesFromFile", prefixesFromFile).
+		WithField("previousPrefixesInfo", prefixesInfo).
+		Debugf("adding excluded prefixes to connection")
+
+	store(ctx, prefixesInfo)
 
```

Comment:
it starts to look more and more complicated.

I think we should consider the problem on the client(nsc) side within exludeprefix.Client chain element that will simply translate response prefixes to client prefixes. In this case all applications in the chain of request can always append prefixes and don't care about diff/conflict cases.

Reply:
ok, it sounds like a good solution, added it